components/codec/number.rs

```diff
@@ -0,0 +1,2096 @@
+    /// The original implementation in TiKV
+    fn original_decode_u64_le(data: &mut &[u8]) -> super::Result<u64> {
+        use byteorder::ByteOrder;
+        read_num_bytes(
+            ::std::mem::size_of::<u64>(),
+            data,
+            byteorder::LittleEndian::read_u64,
+        )
+    }
+
+    /// Decode u64 little endian using `bytes::Buf` over a `Cursor<&[u8]>`.
+    #[bench]
+    fn bench_decode_u64_le_bytes_buf(b: &mut test::Bencher) {
+        use bytes::Buf;
+
+        let buf: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+        b.iter(|| {
+            let mut cursor = ::std::io::Cursor::new(test::black_box(buf.as_slice()));
+            let v = cursor.get_u64_le();
+            assert_eq!(v, 0x0807060504030201);
+            assert_eq!(cursor.position(), 8);
+        });
+    }
+
+    /// Decode u64 little endian using `BufferNumberDecoder` over a `Cursor<&[u8]>`.
+    #[bench]
+    fn bench_decode_u64_le_buffer_decoder(b: &mut test::Bencher) {
+        use super::BufferNumberDecoder;
+
+        let buf: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+        b.iter(|| {
+            let mut cursor = ::std::io::Cursor::new(test::black_box(buf.as_slice()));
+            let v = cursor.read_u64_le().unwrap();
+            assert_eq!(v, 0x0807060504030201);
+            assert_eq!(cursor.position(), 8);
+        });
+    }
+
+    /// Decode u64 little endian using the original implementation in TiKV.
+    #[bench]
+    fn bench_decode_u64_le_original(b: &mut test::Bencher) {
+        let buf: Vec<u8> = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+        b.iter(|| {
+            let mut slice = test::black_box(buf.as_slice());
+            let v = original_decode_u64_le(&mut slice).unwrap();
+            assert_eq!(v, 0x0807060504030201);
+            test::black_box(&slice);
+        });
+    }
+
+    const VARINT_SAMPLE: u64 = 0xDEADBEEFBAAD;
+    const VARINT_ENCODED: [u8; 7] = [173, 245, 190, 247, 219, 213, 55];
+
+    /// Encode u64 in VarInt using `NumberCodec`.
+    #[bench]
+    fn bench_encode_varint_number_codec(b: &mut test::Bencher) {
+        let mut buf: Vec<u8> = vec![0; super::MAX_VARINT64_LENGTH];
+        b.iter(|| {
+            let mut pos = 0;
+            let bytes = super::NumberCodec::encode_var_u64(
+                test::black_box(&mut buf[..]),
+                test::black_box(VARINT_SAMPLE),
+            );
+            pos += bytes;
+            test::black_box(&buf);
+            assert_eq!(pos, VARINT_ENCODED.len());
+        });
+    }
+
+    trait OldVarIntEncoder: byteorder::WriteBytesExt {
+        fn encode_var_u64(&mut self, mut v: u64) -> super::Result<()> {
+            while v >= 0x80 {
+                self.write_u8(v as u8 | 0x80)
+                    .map_err(|_| super::Error::BufferTooSmall)?;
+                v >>= 7;
+            }
+            self.write_u8(v as u8)
+                .map_err(|_| super::Error::BufferTooSmall)
+        }
+    }
+
+    impl<T: byteorder::WriteBytesExt>
```

Comment:
I think you can still benchmark how it performs when the encoded length is 5 or smaller.